File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    diameter=-1
    def maxHeight(self,root):
        if root is None:
            return 0
        
        leftHeight=self.maxHeight(root.left)
        rightHeight=self.maxHeight(root.right)

        self.diameter=max(self.diameter,leftHeight+rightHeight+1)
        return max(leftHeight,rightHeight)+1

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxHeight(root)
        return self.diameter-1

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int noOfNodesInDia=-1;
    int maxHeight(TreeNode* root){
        if(root==NULL)
            return 0;

        int leftHeight=maxHeight(root->left);
        int rightHeight=maxHeight(root->right);

        noOfNodesInDia=max(noOfNodesInDia,leftHeight+rightHeight+1);
        return max(leftHeight,rightHeight)+1;

    }
    int diameterOfBinaryTree(TreeNode* root) {
        maxHeight(root);
        return noOfNodesInDia-1;
    }
};